### PR TITLE
NODE-1525: allow all public API to be used by callback or return a promise

### PR DESCRIPTION
**Pull request by mbroadst**
@

### CODE REVIEW

File: lib/kerberos.js [hunk 1]
```diff
@@ -36,6 +38,93 @@ function validateParameter(parameter, spec) {
...
+    params.push(callback);
+    fn.apply(this, params);
+  };
+}
+
+/**
+ * Processes a single GSSAPI client-side step using the supplied server data.
+ *
+ * @param {string} challenge A string containing the base64-encoded server data (which may be empty for the first step)
+ * @param {function} callback Returns a result code, or an error if one was
```
Comment (kvwalker) [inline lib/kerberos.js:92]: Maybe rephrase "or an error if one was" as "or an error if one occurred"

**Review by kvwalker — APPROVED** [review]
just one nit

File: lib/kerberos.js (hunk 1 above)
Comment (mbroadst) [reply lib/kerberos.js:92]: yeah I think documentation is a bit incomplete, that's the next step 😄

File: lib/kerberos.js [hunk 2]
```diff
@@ -36,6 +38,93 @@ function validateParameter(parameter, spec) {
   }
 }
 
+/// allows api methods to be called with a callback or a promise
```
Comment (daprahamian) [inline lib/kerberos.js]: Maybe add a more verbose description here that also explains the parameter sanitization / schema-ing that you are doing here.

File: lib/kerberos.js [hunk 3]
```diff
@@ -36,6 +38,93 @@ function validateParameter(parameter, spec) {
...
+function promisify(fn, paramDefs) {
```
Comment (daprahamian) [inline lib/kerberos.js]: I think that the name of this should be something other than [`promisify`](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original).

Maybe `defineOperation`?

File: lib/kerberos.js [hunk 4]
```diff
@@ -36,6 +38,93 @@ function validateParameter(parameter, spec) {
...
+  return function() {
+    const args = Array.prototype.slice.call(arguments);
+    const params = [];
+    for (let i = 0; i < paramDefs.length; ++i) {
```
Comment (daprahamian) [inline lib/kerberos.js]: <3 `++i`. Down with `i++`!!!

I feel like I need to make one of those galaxy brain memes for this.

Comment (daprahamian) [reply lib/kerberos.js]: Though actually, it looks like you have a 1 to 1 mapping of `paramDefs` and `params`. Maybe use `.map`?

**Review by daprahamian — CHANGES_REQUESTED** [review]
